File: android/src/main/java/app/capgo/capacitor/compass/CapgoCompassPlugin.java
package app.capgo.capacitor.compass;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CapgoCompass")
public class CapgoCompassPlugin extends Plugin {

    private final String pluginVersion = "8.1.1";
    private CapgoCompass implementation;
    private boolean isListening = false;

    @Override
    public void load() {
        this.implementation = new CapgoCompass(getActivity());
    }

    @Override
    public void handleOnResume() {
        super.handleOnResume();
        if (this.implementation != null && isListening) {
            this.implementation.registerListeners();
        }
    }

    @Override
    public void handleOnPause() {
        super.handleOnPause();
        if (this.implementation != null) {
            this.implementation.unregisterListeners();
        }
    }

    @PluginMethod
    public void getCurrentHeading(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("value", implementation.getCurrentHeading());
        call.resolve(ret);
    }

    @PluginMethod
    public void getPluginVersion(final PluginCall call) {
        try {
            final JSObject ret = new JSObject();
            ret.put("version", this.pluginVersion);
            call.resolve(ret);
        } catch (final Exception e) {
            call.reject("Could not get plugin version", e);
        }
    }

    @PluginMethod
    public void startListening(PluginCall call) {
        if (isListening) {
            call.resolve();
            return;
        }

        isListening = true;
        implementation.setHeadingCallback((heading) -> {
            JSObject ret = new JSObject();
            ret.put("value", heading);
            notifyListeners("headingChange", ret);
        });

        implementation.registerListeners();
        call.resolve();
    }

    @PluginMethod
    public void stopListening(PluginCall call) {
        if (!isListening) {
            call.resolve();
            return;
        }

        isListening = false;
        implementation.setHeadingCallback(null);
        implementation.unregisterListeners();

        call.resolve();
    }

    @PluginMethod
    public void checkPermissions(PluginCall call) {
        // Android does not require any permissions for compass/sensor access
        JSObject ret = new JSObject();
        ret.put("compass", "granted");
        call.resolve(ret);
    }

    @PluginMethod
    public void requestPermissions(PluginCall call) {
        // Android does not require any permissions for compass/sensor access
        JSObject ret = new JSObject();
        ret.put("compass", "granted");
        call.resolve(ret);
    }
}

File: ios/Sources/CapgoCompassPlugin/CapgoCompassPlugin.swift
import CoreLocation
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapgoCompassPlugin)
public class CapgoCompassPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion: String = "7.1.1"
    public let identifier = "CapgoCompassPlugin"
    public let jsName = "CapgoCompass"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getCurrentHeading", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startListening", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopListening", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = CapgoCompass()
    private var isListening = false
    private var permissionCallId: String?

    override public func load() {
        implementation.startListeners()

        NotificationCenter.default.addObserver(
            forName: UIApplication.didBecomeActiveNotification,
            object: nil,
            queue: OperationQueue.main
        ) { [weak self] _ in
            self?.implementation.startListeners()
        }

        NotificationCenter.default.addObserver(
            forName: UIApplication.willResignActiveNotification,
            object: nil,
            queue: OperationQueue.main
        ) { [weak self] _ in
            self?.implementation.stopListeners()
        }
    }

    @objc func getCurrentHeading(_ call: CAPPluginCall) {
        let heading = implementation.getCurrentHeading()
        if heading < 0 {
            call.reject("Failed to get heading. Did the user permit location access?")
            return
        }
        call.resolve([
            "value": heading
        ])
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": self.pluginVersion])
    }

    @objc func startListening(_ call: CAPPluginCall) {
        if isListening {
            call.resolve()
            return
        }

        // Parse optional throttling configuration
        let minInterval = call.getInt("minInterval")
        let minHeadingChange = call.getDouble("minHeadingChange")

        let intervalMs = minInterval ?? 100
        let headingChange = minHeadingChange ?? 2.0
        implementation.setThrottling(minIntervalMs: intervalMs, minHeadingChange: headingChange)

        isListening = true
        implementation.setHeadingCallback { [weak self] heading in
            guard let self = self else { return }
            if heading >= 0 {
                self.notifyListeners("headingChange", data: [
                    "value": heading
                ])
            }
        }

        call.resolve()
    }

    @objc func stopListening(_ call: CAPPluginCall) {
        if !isListening {
            call.resolve()
            return
        }

        isListening = false
        implementation.setHeadingCallback(nil)

        call.resolve()
    }

    @objc override public func checkPermissions(_ call: CAPPluginCall) {
        call.resolve(["compass": currentPermissionState()])
    }

    @objc override public func requestPermissions(_ call: CAPPluginCall) {
        let status = implementation.getAuthorizationStatus()

        if status == .notDetermined {
            // Store call ID and request permission
            permissionCallId = call.callbackId
            implementation.requestPermission { [weak self] in
                guard let self = self else { return }
                if let callId = self.permissionCallId,
                   let savedCall = self.bridge?.savedCall(withID: callId) {
                    savedCall.resolve(["compass": self.currentPermissionState()])
                    self.bridge?.releaseCall(savedCall)
                    self.permissionCallId = nil
                }
            }
            bridge?.saveCall(call)
        } else {
            call.resolve(["compass": currentPermissionState()])
        }
    }

    private func currentPermissionState() -> String {
        let status = implementation.getAuthorizationStatus()
        switch status {
        case .authorizedAlways, .authorizedWhenInUse:
            return "granted"
        case .denied, .restricted:
            return "denied"
        case .notDetermined:
            return "prompt"
        @unknown default:
            return "prompt"
        }
    }
}
